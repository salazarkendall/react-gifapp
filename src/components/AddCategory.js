import React, { useState } from 'react';
import PropTypes from 'prop-types';

/**
 * Este es el Functional Component encargado del funcionamiento del input
 */
export const AddCategory = ({ setCategories }) => {
	/**
	 * Lo siguiente es un hook comun, encargado de la funcionalidad del input
	 */
	const [inputValue, setInputValue] = useState('');

	const handleInputChange = (e) => {
		setInputValue(e.target.value);
	};
	const handleSubmit = (e) => {
		e.preventDefault();

		if (inputValue.trim().length > 2) {
			setCategories((categories) => [...categories, inputValue]);
			setInputValue('');
		}
	};

	return (
		<form onSubmit={handleSubmit}>
			<input
				type="text"
				value={inputValue}
				onChange={handleInputChange}
			/>
		</form>
	);
};

AddCategory.propTypes = {
	setCategories: PropTypes.func.isRequired,
};
