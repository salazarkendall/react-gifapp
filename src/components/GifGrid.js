import React from 'react';

export const GifGrid = ({ category }) => {
	const getGifs = async () => {
		const url = `https://api.giphy.com/v1/gifs/search?q=Dark&limit=10&api_key=GGTJKrckkPhdSGdOc5ZFbTgpFCrcuuNp`;
		const respuesta = await fetch(url);
		const { data } = await respuesta.json();

		const gifs = data.map((imagen) => {
			return {
				id: imagen.id,
				title: imagen.title,
				url: imagen.images?.downsized_medium.url,
			};
		});

		console.log(gifs);
	};

	getGifs();

	return (
		<>
			<h3>{category}</h3>
		</>
	);
};
