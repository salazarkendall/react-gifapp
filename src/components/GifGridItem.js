import React from 'react';
import PropTypes from 'prop-types';
/**
 * Se encarga de darle la estructura HTML a cada uno de los gifs que se mostraran en pantalla
 * Los unicos valores recibidor por parametro son title y url porque asi fue especificado al usar el operador spread
 */
export const GifGridItem = ({ title, url }) => {
    return (
        <div className="card animate__animated animate__zoomIn">
            <img src={url} alt={title} />
            <p>{title}</p>
        </div>
    );
};

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
};
