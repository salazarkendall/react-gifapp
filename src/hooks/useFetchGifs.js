import { useEffect, useState } from 'react';
import { getGifs } from '../helpers/getGifs';

/**
 * Este es nuestro custom hook: useFetchGifs, como funciona?
 * 1. Recibe una categoria de la busqueda por gifs
 * 2. Nuestro custom hook tiene un hook normal, cuyo useState va a contener la data y el loading
 * 3. Dentro de este tambien encontramos un useEffect, que tambien es un hook, que se encarga de:
 * 		3.a. Tomar la categoria recibida por parametros
 * 		3.b. Mandarla al getGifs, la cual es una funcion que realiza una peticion HTTP al API de giphy
 * 		3.c. El return se almacena dentro del hook state, en la parte de data, ademas de que se cambia el loading a false
 * 		3.d. Recordemos que los useEffect solo se llaman 1 vez y tienen dependencias
 * 4. Por ultimo, el state es el valor que se retorna, el cual contiene la data y el loading
 */
export const useFetchGifs = (category) => {
	/**
	 * El state puede contener cualquier cosa, desde un string vacio hasta un objeto como en este caso
	 */
	const [state, setState] = useState({
		data: [],
		loading: true,
	});

	/**
	 * Normalmente las peticiones tardan en realizarse, al menos unos cuantos segundos
	 * En este caso el setTimeout es innecesario y solo sirve para fines ilustrativos
	 */
	useEffect(() => {
		getGifs(category).then((gifs) => {
			setTimeout(() => {
				setState({
					data: gifs,
					loading: false,
				});
			}, 1000);
		});
	}, [category]);

	/**
	 * Este valor de retorno contiene toda la informacion de los nuevos gifs que seran mostrados en pantalla
	 */
	return state;
};
