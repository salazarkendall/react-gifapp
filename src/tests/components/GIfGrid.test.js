import React from 'react';
import '@testing-library/jest-dom';

import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en GifGrid', () => {
	const category = 'One Punch';

	test('Debe de mostrar el componente correctamente', () => {
		useFetchGifs.mockReturnValue({
			data: [],
			loading: true,
		});

		const wrapper = shallow(<GifGrid category={category} />);
		expect(wrapper).toMatchSnapshot();
	});

	test('Debe de mostrar items cuando se cargan imagenes', () => {
		const gifs = [
			{
				id: 'ABC',
				url: 'https://localhost/cualquier/cosa.jpg',
				title: 'Cualquier cosa',
			},
			{
				id: '123',
				url: 'https://localhost/cualquier/cosa.jpg',
				title: 'Cualquier cosa',
			},
			{
				id: '321',
				url: 'https://localhost/cualquier/cosa.jpg',
				title: 'Cualquier cosa',
			},
		];

		useFetchGifs.mockReturnValue({
			data: gifs,
			loading: false,
		});

		const wrapper = shallow(<GifGrid category={category} />);

		// No vamos a evaluar con el snapshot, porque seria una prueba dudosa
		expect(wrapper.find('p').exists()).toBe(false);
		expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
	});
});
