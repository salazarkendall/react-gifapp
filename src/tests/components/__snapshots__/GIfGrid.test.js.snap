// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifGrid Debe de mostrar el componente correctamente 1`] = `
<Fragment>
  <h3>
    One Punch
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
